### PR TITLE
GCS offload support(3): add configs to support GCS driver

### PR DESCRIPTION
**Pull request by zhaijack**
This is the third part to support Google Cloud Storage offload.
It aims to support GCS related config in `ManagedLedgerOffloader`.  It is based on PR #2065.  Please only review the commits after [e5b1f7a](https://github.com/apache/incubator-pulsar/pull/2065/commits/e5b1f7aad4fc1f0d8f4c61b69806100a38774633).

Currently it passed the real test in GCS, by test case [ManagedLedgerOffloaderTest#testGcsRealOffload](https://github.com/apache/incubator-pulsar/pull/2151/commits/eda5097f110acfd3fc5ee518dd8bc16771707c27#diff-6387a2ab4cf9c9243135b5a34c8522efR584), since lack of GCS mock docker image, will try to use read GCS to do the integration test in later PR.

Master Issue: #2067

### CODE REVIEW

**Comment by zhaijack** [comment]
retest this please

File: conf/broker.conf [hunk 1]
```diff
@@ -507,6 +509,24 @@ s3ManagedLedgerOffloadMaxBlockSizeInBytes=67108864
...
 s3ManagedLedgerOffloadReadBufferSizeInBytes=1048576
 
+# For Google Cloud Storage ledger offload region
+# reference this page for more details: https://cloud.google.com/storage/docs/bucket-locations
+gcsManagedLedgerOffloadRegion=
+
+# For Google Cloud Storage ledger offload, Bucket to place offloaded ledger into
+gcsManagedLedgerOffloadBucket=
+
+# For Google Cloud Storage ledger offload, Max block size in bytes. (64MB by default, 5MB minimum)
```
Comment (ivankelly) [inline conf/broker.conf]: 5MB minimum was an AWS imposed constraint. It's probably different for GCS.

File: conf/broker.conf [hunk 2]
```diff
@@ -507,6 +509,24 @@ s3ManagedLedgerOffloadMaxBlockSizeInBytes=67108864
...
+gcsManagedLedgerOffloadMaxBlockSizeInBytes=67108864
+
+# For Google Cloud Storage ledger offload, Read buffer size in bytes (1MB by default)
+gcsManagedLedgerOffloadReadBufferSizeInBytes=1048576
+
+# Note: you need first config this to get gcs avliable.
```
Comment (ivankelly) [inline conf/broker.conf]: Remove this line. It's redundant.

File: conf/broker.conf [hunk 3]
```diff
@@ -507,6 +509,24 @@ s3ManagedLedgerOffloadMaxBlockSizeInBytes=67108864
...
+# For Google Cloud Storage credentials of service account key path, usually it is a json file.
```
Comment (ivankelly) [inline conf/broker.conf]: For Google Cloud Storage, path to json file containing service account credentials.
For more details, see the "Service Accounts" section of https://support.google.com/googleapi/answer/6158849

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/ServiceConfiguration.java [hunk 4]
```diff
@@ -502,6 +502,24 @@
...
+    // For Google Cloud Storage ledger offload, Max block size in bytes.
+    @FieldContext(minValue = 5242880) // 5MB
+    private int gcsManagedLedgerOffloadMaxBlockSizeInBytes = 64 * 1024 * 1024; // 64MB
+
+    // For Google Cloud Storage ledger offload, Read buffer size in bytes.
+    @FieldContext(minValue = 1024)
+    private int gcsManagedLedgerOffloadReadBufferSizeInBytes = 1024 * 1024; // 1MB
+
+    // For Google Cloud Storage credentials of service account key path
+    // reference this page for more details of service account key: https://support.google.com/googleapi/answer/6158849
```
Comment (ivankelly) [inline pulsar-broker-common/src/main/java/org/apache/pulsar/broker/ServiceConfiguration.java]: All comments from conf file also apply here, though I'm not sure if this comments are even useful. We should probably just group them all together and point user to broker.conf if they want details, as they will never update the values in this file.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/offload/impl/BlobStoreManagedLedgerOffloader.java [hunk 5]
```diff
@@ -104,30 +119,61 @@ public static BlobStoreManagedLedgerOffloader create(ServiceConfiguration conf,
...
+
+        if (isGcsDriver(driver) && Strings.isNullOrEmpty(gcsKeyPath)) {
+            throw new PulsarServerException(
+                "The service account key path is empty for GCS driver");
+        }
+        if (isGcsDriver(driver)) {
+            try {
+                gcsKeyContent = Files.toString(new File(gcsKeyPath), Charset.defaultCharset());
+            } catch (IOException ioe) {
+                log.error("meet exception when read gcs service account key file ");
```
Comment (ivankelly) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/offload/impl/BlobStoreManagedLedgerOffloader.java]: -> "Cannot read GCS service account credentials file " + gcsKeyPath

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/offload/impl/BlobStoreManagedLedgerOffloader.java [hunk 6]
```diff
@@ -143,23 +189,35 @@ public static BlobStoreManagedLedgerOffloader create(ServiceConfiguration conf,
...
+        // credentials:
+        //   for s3, get by DefaultAWSCredentialsProviderChain.
+        //   for gcs, use downloaded file 'google_creds.json', which contains service account key by
+        //     following instructions in page https://support.google.com/googleapi/answer/6158849
         String id = "accesskey";
         String key = "secretkey";
-        if (credentials != null) {
-            id = credentials.getAWSAccessKeyId();
-            key = credentials.getAWSSecretKey();
+        if (isS3Driver(driver)) {
```
Comment (ivankelly) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/offload/impl/BlobStoreManagedLedgerOffloader.java]: We shouldn't have AWS/GCS specific logic in this class. Instead, it should be in the factory method, maybe broken out into a utility method called from the factory method. Then the id & key should be passed in as parameters to the constructor of this class.

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/offload/impl/BlobStoreManagedLedgerOffloaderTest.java [hunk 7]
```diff
@@ -179,6 +180,56 @@ public void testSmallBlockSizeConfigured() throws Exception {
...
+            // correct
+            log.error("expect pse", pse);
+        }
+    }
+
+    @Test
+    public void testGcsNoBucketConfigured() throws Exception {
+        ServiceConfiguration conf = new ServiceConfiguration();
+        conf.setManagedLedgerOffloadDriver("google-cloud-storage");
+        //conf.setGcsManagedLedgerOffloadServiceAccountKeyFile("~/Downloads/project-804d5e6a6f33.json");
```
Comment (ivankelly) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/offload/impl/BlobStoreManagedLedgerOffloaderTest.java]: Don't commit commented out code.

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/offload/impl/BlobStoreManagedLedgerOffloaderTest.java [hunk 8]
```diff
@@ -179,6 +180,56 @@ public void testSmallBlockSizeConfigured() throws Exception {
...
+        conf.setGcsManagedLedgerOffloadServiceAccountKeyFile(tmpKeyFile.getAbsolutePath());
+        conf.setGcsManagedLedgerOffloadBucket(BUCKET);
+        conf.setGcsManagedLedgerOffloadMaxBlockSizeInBytes(1024);
+
+        try {
+            BlobStoreManagedLedgerOffloader.create(conf, scheduler);
+            Assert.fail("Should have thrown exception");
+        } catch (PulsarServerException pse) {
+            // correct
+            log.error("expect pse", pse);
```
Comment (ivankelly) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/offload/impl/BlobStoreManagedLedgerOffloaderTest.java]: log.info("Expected pse", pse);

File: conf/broker.conf [hunk 9]
```diff
@@ -507,6 +509,24 @@ s3ManagedLedgerOffloadMaxBlockSizeInBytes=67108864
 # For Amazon S3 ledger offload, Read buffer size in bytes (1MB by default)
 s3ManagedLedgerOffloadReadBufferSizeInBytes=1048576
 
+# For Google Cloud Storage ledger offload region
```
[the remark this answers is not in the document]
Comment (zhaijack) [reply conf/broker.conf]: Thanks, Yes, It is also 5 MB in GCS. 
https://cloud.google.com/storage/docs/json_api/v1/how-tos/multipart-upload
Seems GCS make some alignment with aws s3

**Comment by zhaijack** [comment]
@ivankelly Thanks for the comments, updated it.

**Pull request merged**